sdks/python/apache_beam/transforms/environments.py

```diff
@@ -125,9 +125,17 @@ def __init__(self,
         dict(resource_hints) if resource_hints else {})
 
   def __eq__(self, other):
+    equal_artifacts = True
+    for first, second in zip(self._artifacts, other._artifacts):
+      # do not compare type payload since it contains a unique hash.
```

Comment:
This comparison is brittle because:

1) we rely on a particular ordering of self._artifacts. It sounds like the reason you are changing it is that you have two essentially equal environments that include an artifact with identical payload/sha256 but different source file location. The source file location is affected by temp folder that introduces randomness. This randomness will be reflected in `artifact.SerializeToString()`, which is used as a key to `sort self._artifacts` (line 121). Because of that, there is no guarantee that when you iterate over `self._artifacts` and `other._artifacts`, multiple artifacts will be ordered in a consistent manner.
2) ignoring entire type.payload might backfire. I would rather use a different key for staged files, than `x.SerializeToString()`. For local files, it could be evaluated based on sha256 and destination name but ignore the source location. Then, you can use such key when sorting artifacts, and when comparing them. Having such key would mean that we'd have to decode ArtifactInformation, and potentially compare different artifact types/roles differently. We could still fall back to `x.SerializeToString()`.

Alternatively you could adjust requirement generation to use the same temp directory for generating the pip freeze over the course of the python process or the test. I think we could do something to prevent the issue you are dealing with here.

Reply:
with same temp directory, won't it overwrite the file with same name? because the two values are like this:
```
first: [type_urn: "beam:artifact:type:file:v1"
type_payload: "\nd/var/folders/kk/jk392txd7x14sg93r8sjnzkh00tp20/T/tmpmxw3v4d9/submission_environment_dependencies.txt"
role_urn: "beam:artifact:role:staging_to:v1"
role_payload: "\n\'submission_environment_dependencies.txt"
]
second: [type_urn: "beam:artifact:type:file:v1"
type_payload: "\nd/var/folders/kk/jk392txd7x14sg93r8sjnzkh00tp20/T/tmpni8k_038/submission_environment_dependencies.txt"
role_urn: "beam:artifact:role:staging_to:v1"
role_payload: "\n\'submission_environment_dependencies.txt"
]
```

But yeah I agree that we need to fix the artifact comparison function, somehow this wasn't caught earlier!

Reply:
Okay, so I reverted this back to how it was originally. creating a mock for the tests that writes artifacts to the same directory resolves this issue.